### PR TITLE
DbActor: fill decodedData

### PR DESCRIPTION
**Pull request by lourkeur**
Whenever a message is retrieved from the db through the `DbActor` (e.g. by sending a `Read` or `Catchup` signal), the `decodedData` field is not filled.

### CODE REVIEW

File: be2-scala/src/main/scala/ch/epfl/pop/storage/DbActor.scala [hunk 1]
```diff
@@ -47,7 +48,11 @@ final case class DbActor(
   @throws [DbActorNAckException]
   private def read(channel: Channel, messageId: Hash): Option[Message] = {
     Try(storage.read(s"$channel${Channel.DATA_SEPARATOR}$messageId")) match {
-      case Success(Some(json)) => Some(Message.buildFromJson(json))
+      case Success(Some(json)) =>
+        val msg = Message.buildFromJson(json)
+        val channelData = readChannelData(channel)
+        val builder = registry.getBuilder(channelData.channelType, ActionType.CREATE).get
```
Comment (lourkeur) [inline be2-scala/src/main/scala/ch/epfl/pop/storage/DbActor.scala:54]: I'm not sure how to properly summon the correct action type...

**Comment by lourkeur** [comment]
If we want a proper `(JsonString, MessageRegistry) => MessageData` procedure to be possible, the `MessageData` json schema has to include these fields...

**Comment by lourkeur** [comment]
As far as the current state of the PR is concerned: the plan would be to replace all the `jsonFormatN` calls in `MessageDataProtocol` like I have already done 3 times. It just requires more time to sit down and do. (or maybe we can meta-program our way out of that)

**Review by Ajkunas — DISMISSED** [review]
Tested!

File: be2-scala/src/main/scala/ch/epfl/pop/json/MessageDataProtocol.scala [hunk 2]
```diff
@@ -217,25 +244,25 @@ object MessageDataProtocol extends DefaultJsonProtocol {
     }
   }
 
-  implicit val closeRollCallFormat: JsonFormat[CloseRollCall] = jsonFormat[Hash, Hash, Timestamp, List[PublicKey], CloseRollCall](CloseRollCall.apply, "update_id", "closes", "closed_at", "attendees")
-  implicit val createRollCallFormat: JsonFormat[CreateRollCall] = jsonFormat[Hash, String, Timestamp, Timestamp, Timestamp, String, Option[String], CreateRollCall](CreateRollCall.apply, "id", "name", "creation", "proposed_start", "proposed_end", "location", "roll_call_description")
-  implicit val openRollCallFormat: JsonFormat[OpenRollCall] = jsonFormat[Hash, Hash, Timestamp, OpenRollCall](OpenRollCall.apply, "update_id", "opens", "opened_at")
-  implicit val reopenRollCallFormat: JsonFormat[ReopenRollCall] = jsonFormat[Hash, Hash, Timestamp, ReopenRollCall](ReopenRollCall.apply, "update_id", "opens", "opened_at")
+  implicit val closeRollCallFormat: JsonFormat[CloseRollCall] = annotateHeader(jsonFormat[Hash, Hash, Timestamp, List[PublicKey], CloseRollCall](CloseRollCall.apply, "update_id", "closes", "closed_at", "attendees"))
```
Comment (tuomas-p) [inline be2-scala/src/main/scala/ch/epfl/pop/json/MessageDataProtocol.scala:247]: the new formats look good to me

File: be2-scala/src/main/scala/ch/epfl/pop/json/MessageDataProtocol.scala [hunk 3]
```diff
@@ -37,8 +41,31 @@ object MessageDataProtocol extends DefaultJsonProtocol {
...
 
+  // ------------------------------- METADATA UTILITY -------------------------------------- //
+  // retrieve information from the header if it is correct.
+  //
+  // Succeeds if both 'object' and 'action' are present and are strings
+  def parseHeader(data: String): Try[(ObjectType, ActionType)] =
+    Try {
+      data.parseJson.asJsObject.getFields(PARAM_OBJECT, PARAM_ACTION) match {
+        case Seq(objectString @ JsString(_), actionString @ JsString(_)) =>
+          (objectString.convertTo[ObjectType], actionString.convertTo[ActionType])
```
Comment (tuomas-p) [inline be2-scala/src/main/scala/ch/epfl/pop/json/MessageDataProtocol.scala:52]: I don't exactly know whether this is really a relevant point, as it's wrapped in the Try, but I'd still add a case _ => just to be sure

**Review by tuomas-p — CHANGES_REQUESTED** [review]
Apart from the small change with the match case (which I'm not completely sure about), looks good to me, very clean overall!

File: be2-scala/src/main/scala/ch/epfl/pop/json/MessageDataProtocol.scala (hunk 3 above)
Comment (lourkeur) [reply be2-scala/src/main/scala/ch/epfl/pop/json/MessageDataProtocol.scala:52]: That is absolutely right, fixed it up!

File: be2-scala/src/main/scala/ch/epfl/pop/json/MessageDataProtocol.scala [hunk 4]
```diff
@@ -37,8 +41,32 @@ object MessageDataProtocol extends DefaultJsonProtocol {
...
+        case _ => throw new IllegalArgumentException("parseHeader: header fields not found")
```
Comment (tuomas-p) [inline be2-scala/src/main/scala/ch/epfl/pop/json/MessageDataProtocol.scala:53]: 👍

**Review by tuomas-p — APPROVED** [review]

**Pull request merged**